trieste/ask_tell_optimization.py

```diff
@@ -0,0 +1,290 @@
+            - ``datasets`` or ``model_specs`` are empty
+            - default acquisition is used but incompatible with other inputs
+        """
+        self._search_space = search_space
+        self._acquisition_state = acquisition_state
+
+        if not datasets or not model_specs:
+            raise ValueError("dicts of datasets and model_specs must be populated.")
+
+        if isinstance(datasets, Dataset):
+            datasets = {OBJECTIVE: datasets}
+            # ignore below is due to MyPy not being able to handle overloads properly
+            model_specs = {OBJECTIVE: model_specs}  # type: ignore
+
+        # reassure the type checker that everything is tagged
+        datasets = cast(Dict[str, Dataset], datasets)
+        model_specs = cast(Dict[str, ModelSpec], model_specs)
+
+        if datasets.keys() != model_specs.keys():
+            raise ValueError(
+                f"datasets and model_specs should contain the same keys. Got {datasets.keys()} and"
+                f" {model_specs.keys()} respectively."
+            )
+
+        self._datasets = datasets
+        self._models = map_values(create_model, model_specs)
+
+        if acquisition_rule is None:
+            if self._datasets.keys() != {OBJECTIVE}:
+                raise ValueError(
+                    f"Default acquisition rule EfficientGlobalOptimization requires tag"
+                    f" {OBJECTIVE!r}, got keys {self._datasets.keys()}"
+                )
+
+            self._acquisition_rule = cast(
+                AcquisitionRule[TensorType, SP], EfficientGlobalOptimization()
+            )
+        else:
+            self._acquisition_rule = acquisition_rule
+
+        if fit_model:
+            for tag, model in self._models.items():
+                dataset = datasets[tag]
+                model.update(dataset)
+                model.optimize(dataset)
+
+    def __repr__(self) -> str:
+        """Print-friendly string representation"""
+        return f"""AskTellOptimizer({self._search_space!r}, {self._datasets!r},
+               {self._models!r}, {self._acquisition_rule!r}), "
+               {self._acquisition_state!r}"""
+
+    @classmethod
+    def from_record(
+        cls,
+        record: Record[S],
+        search_space: SP,
+        acquisition_rule: AcquisitionRule[TensorType | State[S | None, TensorType], SP]
+        | None = None,
+    ) -> AskTellOptimizer[SP]:
+        """Creates new :class:`~AskTellOptimizer` instance from provided optimization state.
+
+        :param record: Optimization state record.
+        :param search_space: The space over which to search for the next query point.
+        :param acquisition_rule: The acquisition rule, which defines how to search for a new point
+            on each optimization step. Defaults to
+            :class:`~trieste.acquisition.rule.EfficientGlobalOptimization` with default
+            arguments.
+        """
+        # we are recovering previously saved optimization state
```

Comment:
perhaps move this comment to the docstring?

Reply:
Left it where it is, but added some short explanation of this behavior to the docstring too